lib/phase/flatten.js

```diff
@@ -19,41 +42,91 @@ function flattenSharedIntoJobs(shared, jobs) {
         const oldJob = clone(jobs[jobName]);
 
         // Replace
-        ['image', 'matrix', 'steps'].forEach((key) => {
+        ['image', 'matrix', 'steps', 'template'].forEach((key) => {
             if (oldJob[key]) {
                 newJob[key] = oldJob[key];
             }
         });
 
-        if (!newJob.environment) {
-            newJob.environment = {};
-        }
-        if (!newJob.secrets) {
-            newJob.secrets = [];
-        }
-        if (!newJob.settings) {
-            newJob.settings = {};
-        }
-
-        // Merge
-        Object.assign(newJob.environment, oldJob.environment || {});
-        Object.assign(newJob.settings, oldJob.settings || {});
-        newJob.secrets = newJob.secrets.concat(oldJob.secrets || []);
-
+        merge(newJob, oldJob);
         newJobs[jobName] = newJob;
     });
 
     return newJobs;
 }
 
 /**
+ * Retrieve template and merge into job config
+ *
+ * @method mergeTemplateIntoJob
+ * @param  {String}           jobName           Job name
+ * @param  {Object}           jobConfig         Job config
+ * @param  {Object}           newJobs           List of new jobs
+ * @param  {TemplateFactory}  templateFactory   Template Factory to get templates
+ * @return {Promise}
+ */
+function mergeTemplateIntoJob(jobName, jobConfig, newJobs, templateFactory) {
+    const matched = TEMPLATE_REGEX.exec(jobConfig.template);
+    const name = matched[NAME_MATCH];
+    const version = matched[VERSION_MATCH];
+    const label = matched[LABEL_MATCH];
+    const oldJob = jobConfig;
+
+    // Try to get the template
+    return templateFactory.getTemplate({ name, version, label })
+        .then((template) => {
+            if (!template) {
+                throw new Error('Template does not exist');
+            }
+
+            const newJob = template.config;
+
+            merge(newJob, oldJob);
+            delete newJob.template;
+            newJobs[jobName] = newJob;
+
+            return null;
+        });
+}
+
+/**
+ * Goes through each job and if template is specified, then merge into job config
+ *
+ * @method flattenTemplates
+ * @param  {Object}   jobs      List of jobs (name => job)
+ * @param  {String}   apiUrl    API Url to retrieve templates from
+ * @return {Promise}            Updated list of jobs after merging templates
+ */
+function flattenTemplates(jobs, apiUrl) {
+    const newJobs = {};
+    const templates = [];
+
+    // eslint-disable-next-line
+    Object.keys(jobs).forEach((jobName) => {
+        const jobConfig = clone(jobs[jobName]);
+        const templateConfig = jobConfig.template;
+
+        // If template is specified, then merge
+        if (templateConfig) {
+            templates.push(mergeTemplateIntoJob(jobName, jobConfig, newJobs, apiUrl));
```

Comment:
The method signature is wrong: `mergeTemplateIntoJob(jobName, jobConfig, newJobs, templateFactory)`